### PR TITLE
Fix bridging authorization

### PR DESCRIPTION
**Pull request by luis4a0**
The prompt asking the user for authorization to disrupt networking to create a bridge was not working. This PR fixes issue 1 of [this review](https://github.com/canonical/multipass/pull/3195#pullrequestreview-1765804396).

### CODE REVIEW

**Review by townsend2010 — CHANGES_REQUESTED** [review]
Hi @luis4a0!  Thanks for this!

It appears the output is not quite correct when it prompts for allowing to create the bridge.  Here is what I see:
```
$ multipass set local.test-issue-1.bridged=true
Multipass needs to create a bridge to connect to true.
This will temporarily disrupt connectivity on that interface.

Do you want to continue (yes/no)?
```
The first line says `Multipass needs to create a bridge to connect to true.` and notice that it ends with `true` which is not correct.

Also, I really feel this is overloading generic commands/RPC in a very specific way which is kind of changing this interface to bend to the needs of asking for bridging authorization.  I feel the asking for bridging authorization should somehow be decoupled somewhat from all of this or at least make it to where the prompting could be more generalized.  I think we need to give this some more thought and I would also like @ricab's opinion on this as well.

**Comment by townsend2010** [comment]
I will add that aside from the wrong string that I pointed out, it does work as intended. :slightly_smiling_face:

**Comment by ricab** [comment]
I agree with @townsend2010. The handling for the bridging prompt is too specific for the generic set code. I would rather see a generic scheme to accommodate generic confirmation prompts, ideally applying to all commands and covering other cases of confirmation/choice prompts (e.g. auto snapshot prompt in restore). That said, I am not sure that this is the right time and place for that... Maybe there's a more accessible middle ground for now? I wouldn't mind seeing a proposal for what shape that could take. WDYT @luis4a0?

**Comment by townsend2010** [comment]
I think a more generic `Prompter` type is better.  Thinking about it some, I can envision a Yes/No prompter, essentially something that accepts a `bool`.  I also envision a prompter that would accept a string.  So yes, for this PR, I think switching to a `YesNoPrompter` that accepts a custom string used for displaying to the user is good.

Also, I'm thinking instead of the "custom" failure handling, using the `StreamingCallback` may be more efficient here.  It's what I used to implement the SMB mount password handling in Windows and seemed to work well for client/daemon ping ponging.

**Comment by luis4a0** [comment]
@townsend2010, fixed the `true` in the prompt. Now the displayed network is correct.

**Review by townsend2010 — CHANGES_REQUESTED** [review]
Hi @luis4a0!

Thanks for the changes you've done.  I have one request below and then I think we are good pending your investigation as to why the test coverage fell so much.

File: src/client/cli/cmd/remote_settings_handler.cpp
Comment (luis4a0) [reply src/client/cli/cmd/remote_settings_handler.cpp]: Hey @townsend2010, thanks for the new review. Fixed.

Comment (townsend2010) [reply src/client/cli/cmd/remote_settings_handler.cpp]: Thanks!

File: src/daemon/daemon.cpp
Comment (luis4a0) [reply src/daemon/daemon.cpp:2586]: There is one way of testing that. But for that, as far as I investigated, I'll have to mock the bridging prompt (or the terminal) and create a mocked daemon with lots of functionality. Do you think adding that coverage is worth the effort?

File: tests/test_daemon.cpp
Comment (townsend2010) [inline tests/test_daemon.cpp:2161]: Should you set an expectation on the status as well?

File: tests/test_instance_settings_handler.cpp
Comment (townsend2010) [inline tests/test_instance_settings_handler.cpp:529]: Not sure I understand why this is here in the test.  It doesn't seem to do anything.

File: tests/test_instance_settings_handler.cpp
Comment (townsend2010) [inline tests/test_instance_settings_handler.cpp:547]: Again, not sure why this is at the end of the test?

File: tests/test_common_callbacks.cpp
Comment (townsend2010) [inline tests/test_common_callbacks.cpp:185]: It's not clear to me what this is actually testing.  Is it just ensuring the client writes *something*?

File: src/daemon/daemon.cpp
Comment (townsend2010) [reply src/daemon/daemon.cpp:2586]: I'm not sure why you need to do all of that.  You basically already have everything in place to test this.  You just need to make sure that the mock `server` when it reads the requests sets `authorized` to `true` and check to make sure the `debug` log message has the expected string.  Nothing about the bridging prompt or terminal needs to be mocked for this.

**Review by townsend2010 — COMMENTED** [review]
Hey @luis4a0,

I answered your reply about more coverage and have some more questions below regarding the tests.

File: src/daemon/daemon.cpp
Comment (townsend2010) [reply src/daemon/daemon.cpp:2586]: Here is a test that will cover this.  It's very similar to the `setWorksWhenBridgingThrows` test.
```c++
TEST_F(Daemon, setWorksAfterAuthorizationRequested)
{
    const std::string key{"local.instance.bridged"}, value{"true"};

    mp::Daemon daemon{config_builder.build()};

    auto logger_scope = mpt::MockLogger::inject();
    logger_scope.mock_logger->screen_logs(mpl::Level::debug);
    logger_scope.mock_logger->expect_log(mpl::Level::debug,
                                         fmt::format("Asking for user authorization to set {}={}", key, value));
    logger_scope.mock_logger->expect_log(mpl::Level::debug, fmt::format("Succeeded setting {}={}", key, value));

    mp::SetRequest request;
    request.set_key(key);
    request.set_val(value);

    const auto& exception = mp::NonAuthorizedBridgeSettingsException{"reason", "instance", "eth8"};

    EXPECT_CALL(mock_settings, set).WillOnce(Throw(exception)).WillOnce(Return());

    auto mock_server = StrictMock<mpt::MockServerReaderWriter<mp::SetReply, mp::SetRequest>>{};

    EXPECT_CALL(mock_server, Write(_, _)).WillOnce([](mp::SetReply reply, auto) {
        EXPECT_TRUE(reply.needs_authorization());
        return true;
    });

    EXPECT_CALL(mock_server, Read(_)).WillOnce([](mp::SetRequest* request) {
        request->set_authorized(true);
        return true;
    });

    EXPECT_TRUE(call_daemon_slot(daemon, &mp::Daemon::set, request, mock_server).ok());
}
```

Comment (luis4a0) [reply src/daemon/daemon.cpp:2586]: Hey @townsend2010, thanks! Yeah, that works indeed. What I did is to add two versions of this test, that is, with positive and negative answers from user. That way, all the code is tested and we don't need the other tests. I removed them.

File: tests/test_daemon.cpp
Comment (luis4a0) [reply tests/test_daemon.cpp]: Removed this test, see comment below.

File: tests/test_instance_settings_handler.cpp
Comment (luis4a0) [reply tests/test_instance_settings_handler.cpp]: Right, just wanted to set `bridged_interface` to its previous value, but this is not needed. Removed.

File: tests/test_instance_settings_handler.cpp
Comment (luis4a0) [reply tests/test_instance_settings_handler.cpp]: As above, I just wanted to set `user_authorized` to its previous value, but again this is not needed. Removed.

File: src/daemon/daemon.cpp
Comment (townsend2010) [reply src/daemon/daemon.cpp:2586]: Excellent, thanks!

**Review by townsend2010 — APPROVED** [review]
Hey @luis4a0!

This looks much better!  Thanks!  I'm approving, but I will wait on CI to complete before adding it to the merge queue.

**Pull request merged**
